Create pydough from_string API (#383)
bodo-ai/PyDough: documentation/usage.md

- [Evaluation APIs](#evaluation-apis)
   * [`pydough.to_sql`](#pydoughto_sql)
   * [`pydough.to_df`](#pydoughto_df)

>    * [`pydough.from_string`](#pydoughfrom_string)

Reviewer: This is not an evaluation API, because it doesn't evaluate anything. This should be its own alternative section after `Setting Up in Jupyter Notebooks`

See the [demo notebooks](../demos/notebooks/1_introduction.ipynb) for more instances of how to use the `to_df` API.

<!-- TOC --><a name="pydoughfrom_string"></a>
### `pydoughfrom_string`

> The `from_string` API parses and transform a PyDough source code string into an unqualified node on which operations like `explain()`, `to_sql()`, or `to_df()` can be called. The first argument is the source code string. It can be a single pydough command or a multi-line pydough code with intermediate results stored in variables. 

Reviewer: Don't say "unqualified node", just say `PyDough code`. This is a user document, and users shouldn't need to know/learn what unqualified nodes are.

Author: Done!

The second argument is the name of the variable that stores the final result of pydough code, for further operations. The API will return the unqualified node containing this final result of the pydough code. If the second argument is not provided, the result will be expected to be in a variable called `result`.

The third argument is optional and corresponds to the pydough graph metadata that needs to be used for the transformation. If not provided, `active_session.metadata` will be used.

> An optional fourth argument allows to specify additional environment context for the pydough code. This environment context will be the local namespace for the executed code. If provided, it should be a dictionary.

Reviewer: For consistency, please reformat this into the same format that all of our other argument docstrings are in, like `to_df` and `to_sql`. E.g. after the first argument, the string, do something like:

```
The `from_string` API can optionally take in the following keyword arguments:

- `answer_variable`: ...
- `metadata`: ...
- `environment`: ...
```

Also some details that should be specified:
- It is assumed that the PyDough code string includes a variable definition where the name of the variable is the same as `answer_variable` and the value is valid PyDough code; if not it raises an exception.
- The fourth argument will be modified in-place when the string is executed as code, as it is where any variables defined in the code will be stored.

Reviewer: Duly noted

> Below are examples of using `pydough.from_string` and the output (the SQL output may be outdated if PyDough's SQL conversion process has been updated):

Reviewer: ```suggestion
Below are examples of using `pydough.from_string`, and examples of the SQL that could be potentially generated from calling `pydough.to_sql` on the output
```

```py
import pydough

> # Setup demo metadata
> graph = pydough.active_session.load_metadata_graph("demos/metadata/tpch_demo_graph.json", "TPCH")
> pydough.active_session.connect_database("sqlite", database="tpch.db")

Reviewer: We don't need the database, and we shouldn't be pointing to a specific graph here.
```suggestion
# Set up the graph
graph = pydough.active_session.load_metadata_graph(...)
```

Reviewer: Don't forget about this

> # Example of a simple pydough code snippet that uses variables provided in the environment context

Reviewer: Split this comment onto two lines

> pydough_code = "result = TPCH.CALCULATE(n=COUNT(customers.WHERE(market_segment == SEG)))"

Author: Done!

# Transform the pydough code and get the UnqualifiedNode with the result
query = pydough.from_string(pydough_code, "result", graph, {"SEG":"AUTOMOBILE"})
sql = pydough.to_sql(query)
print(sql)
```

```sql
SELECT
  COUNT(*) AS n
FROM main.customer
WHERE
  c_mktsegment = 'AUTOMOBILE'
```

> 

Reviewer: Let's put some plain text in front of / between each example so they are clearly delineated.

Reviewer: Don't forget about this

```py

> # Example of a multi-line pydough code snippet that uses intermetiate results to build the final query

Reviewer: Split onto multiple lines

> nation_name = "JAPAN"
> ship_mode = "TRUCK"
> ship_year = 1996
> env = {"NATION_NAME" : nation_name, "SHIP_MODE" : ship_mode, "SHIP_YEAR" : ship_year}

Reviewer: Let's give the env variables different-looking names from the local variables, for clarity
```suggestion
nation_name = "JAPAN"
ship_mode = "TRUCK"
ship_year = 1996
env = {"TARGET_NATION" : nation_name, "DESIRED_SHIP_MODE" : ship_mode, "REQUESTED_SHIP_YEAR" : ship_year}
```

> pydough_code="""

Reviewer: State the goal of this example first.

# The supply records for the supplier that were from a medium part
selected_records = supply_records.WHERE(STARTSWITH(part.name, "coral")).CALCULATE(supply_cost)

# The revenue generated by a specific lineitem
line_revenue = extended_price * (1 - discount) * (1 - tax) - quantity * supply_cost

# The lineitem purchases for each record that were ordered in SHIP_YEAR and shipped via SHIP_MODE
lines_96 = selected_records.lines.WHERE((YEAR(ship_date) == SHIP_YEAR) & (ship_mode == SHIP_MODE)).CALCULATE(rev=line_revenue)

# For each supplier, list their name & selected revenue from SHIP_YEAR
selected_suppliers = suppliers.WHERE(nation.name == NATION_NAME)
supplier_info = selected_suppliers.CALCULATE(name, revenue96=ROUND(SUM(lines_96.rev), 2))

# Pick the 5 suppliers with the highest from SHIP_YEAR
result = supplier_info.TOP_K(5, by=revenue96.DESC())
"""
# Transform the pydough code and get the UnqualifiedNode with the result

> query = pydough.from_string(pydough_code, environment=env)

Reviewer: Let's modify this example to also show using an alternative `answer_variable`.

Reviewer: Don't forget about this

Author: I do it for the first example. It is much simpler.

See the [demo notebooks](../demos/notebooks/1_introduction.ipynb) for more instances of how to use the `to_df` API.

<!-- TOC --><a name="pydoughfrom_string"></a>

> ### `pydoughfrom_string`

Reviewer: ```suggestion
### `pydough.from_string`
```

The `from_string` API parses a PyDough source code text string and and transforms it into a PyDough collection on which operations like `explain()`, `to_sql()`, or `to_df()` can be called. The first argument is the source code string. It can be a single pydough command or a multi-line pydough code with intermediate results stored in variables. 

The second argument is the name of the variable that stores the final result of pydough code, for further operations. The API will return the unqualified node containing this final result of the pydough code. If the second argument is not provided, the result will be expected to be in a variable called `result`.

The third argument is optional and corresponds to the pydough graph metadata that needs to be used for the transformation. If not provided, `active_session.metadata` will be used.

An optional fourth argument allows to specify additional environment context for the pydough code. This environment context will be the local namespace for the executed code. If provided, it should be a dictionary.

> Below are examples of using `pydough.from_string`, and examples of the SQL hat could be potentially generated from calling `pydough.to_sql` on the output.

Reviewer: ```suggestion
Below are examples of using `pydough.from_string`, and examples of the SQL that could be potentially generated from calling `pydough.to_sql` on the output.
```

```

This next example is of Python code to generate SQL to get the top 5 suppliers with the highest revenue. The code snippet uses variables provided in the environment context to filter by nation, ship mode and year (`TARGET_NATION`, `DESIRED_SHIP_MODE` and `REQUESTED_SHIP_YEAR`):
```py
# Example of a multi-line pydough code snippet with intermetiate results
nation_name = "JAPAN"
ship_mode = "TRUCK"
ship_year = 1996
env = {"TARGET_NATION" : nation_name, "DESIRED_SHIP_MODE" : ship_mode, "REQUESTED_SHIP_YEAR" : ship_year}

pydough_code="""
# The supply records for the supplier that were from a medium part
selected_records = supply_records.WHERE(STARTSWITH(part.name, "coral")).CALCULATE(supply_cost)

# The revenue generated by a specific lineitem
line_revenue = extended_price * (1 - discount) * (1 - tax) - quantity * supply_cost

# The lineitem purchases for each record that were ordered in REQUESTED_SHIP_YEAR and shipped via DESIRED_SHIP_MODE
lines_96 = selected_records.lines.WHERE((YEAR(ship_date) == REQUESTED_SHIP_YEAR) & (ship_mode == DESIRED_SHIP_MODE)).CALCULATE(rev=line_revenue)

# For each supplier, list their name & selected revenue from REQUESTED_SHIP_YEAR

> selected_suppliers = suppliers.WHERE(nation.name == TARGET_NATION)

Author: Done!

Below are examples of using `pydough.from_string`, and examples of the SQL that could be potentially generated from calling `pydough.to_sql` on the output. All these examples use the database tcph that can be downloaded running the following command from the root directory of PyDough: `bash demos/setup_tpch.sh tpch.db`
 
Python code using `pydough.from_string` to generate SQL to get the count of customers in the market segment AUTOMOBILE. The result will be returned in a variable named `pydough_query` instead of the default `result`, and the market segment AUTOMOBILE is passed in an environment variable `SEG`.:
```py
import pydough

# Setup demo metadata
# You need to download tcph database using setup_tpch.sh. Check demos/README.md
graph = pydough.active_session.load_metadata_graph("demos/metadata/tpch_demo_graph.json", "TPCH")
pydough.active_session.connect_database("sqlite", database="tpch.db")

# Example of a simple pydough code snippet that uses variables provided in 
# the environment context. This query gets the count of customers from the 
# AUTOMOBILE market segment. For this, a SEG variable is passed in the 

> # environment with a value of "AUTOMOBILE"

Author: We have it outside but not if the code block is copied to run it outside using python.

> # environment with a value of "AUTOMOBILE"

Reviewer: Let's have it in one or the other, not both.

pydough_code = "pydough_query = TPCH.CALCULATE(n=COUNT(customers.WHERE(market_segment == SEG)))"
# Transform the pydough code and get the result from pydough_query
query = pydough.from_string(pydough_code, "pydough_query", graph, {"SEG":"AUTOMOBILE"})
sql = pydough.to_sql(query)
print(sql)
```

The output for `print(sql)` is the following SQL query:
```sql
SELECT
  COUNT(*) AS n
FROM main.customer
WHERE
  c_mktsegment = 'AUTOMOBILE'
```

Python code to generate SQL to get the top 5 suppliers with the highest revenue. The code snippet uses variables provided in the environment context to filter by nation, ship mode and year:
```py
# Example of a multi-line pydough code snippet that uses intermetiate results to
# build the final query. For this query we set 3 environment variables: 
# TARGET_NATION, DESIRED_SHIP_MODE and REQUESTED_SHIP_YEAR. These variables are

> # used in the Pydough source code.

Author: Same. We have it outside but description will be lost if the code block is copied to run it outside using python.

> # used in the Pydough source code.

Author: Done
